### ISSUE TITLE
Update code and docs for generic hosting upcoming changes

### ISSUE THREAD

**Issue by tillig** [autofac/Autofac.Extensions.DependencyInjection#36]
Pull request aspnet/Hosting#1580 includes changes to how hosting works to support a generic host builder on which web host builder is layered. Some of these changes are breaking and we'll need to adjust docs, examples, and code to support it.

- Adding an `IServiceProviderFactory` as a service no longer works. Call `IHostBuilder.UseServiceProviderFactory` instead. This changes the samples and renders the `ServiceCollectionExtensions.AddAutofac` method obsolete.
- You can no longer return an `IServiceProvider` from `Startup.ConfigureServices`. You must use the `IServiceProviderFactory` and `ConfigureContainer`. This will affect the use case where folks intentionally want to keep a reference to the built container or use a child lifetime scope as the pseudo-root of the app. [I've mentioned that in a comment on the PR.](https://github.com/aspnet/Hosting/pull/1580#issuecomment-438372845) You may need a custom service provider factory for this, but you can capture the factory instance during Configure.

In general it appears that while the new pattern moves toward a more generic hosting mechanism, it moves away from the notion of having composite application types (e.g., a self-hosted ASP.NET Core app _and_ a background service all in one logical process).

**Comment by tillig** [comment]
It does appear in [ASP.NET Core 3 the move to the generic host builder will not support this and you'll get an exception](https://github.com/aspnet/AspNetCore/issues/5149) if you try to return your own `IServiceProvider`.

**Comment by tillig** [comment]
Based on #46, it would be nice to see if we can somehow allow the `IServiceProviderFactory` to use a child lifetime scope for the provider it creates rather than assuming the root container. Perhaps some `AddAutofac` overloads would allow that, though I'm not sure at this precise moment what that would look like.

**Comment by tillig** [comment]
Looking at [the `IServiceProviderFactory<TContainerBuilder>` interface](https://github.com/aspnet/Extensions/blob/3ba072c25373c19ab2cfe34483e733f3b926b2c0/src/DependencyInjection/DI.Abstractions/src/IServiceProviderFactory.cs), it assumes that...

- `CreateBuilder` will return a `ContainerBuilder` that has the set of registered services already populated in it - the stuff that came from `Startup.ConfigureServices`.
- `CreateServiceProvider` will take a `ContainerBuilder` _but not necessarily guaranteed to be the same one issued from `CreateBuilder`_ and build a service provider.

In practice, `CreateBuilder` and `CreateServiceProvider` should both get executed from the same provider factory instance since it's registered as a singleton. This is how the default Microsoft provider works as well. It's _likely_ that the `CreateServiceProvider` method will be called with the same `ContainerBuilder` as provided by `CreateBuilder`... but that's not documented, tested, or otherwise guaranteed by the ASP.NET Core framework or the hosting mechanism that I can find. It would be bad if we made that assumption and counted on it since, long term, that assumption could break and we'd be on the hook to try and figure out ways around it.

I'm also not sure how to return the `ContainerBuilder` that's the parameter in a lambda expression that has to run when you build the lifetime scope. Like, when you create a child lifetime scope, it's...

```c#
var scope = container.BeginLifetimeScope(scopeBuilder => scopeBuilder.RegisterType<T>());
```

Like that. It's a lambda expression. There isn't actually a `ContainerBuilder` object. So I don't know what `CreateBuilder` would return, because logically it needs to return _that scope-level builder_. There's also not currently a way to take a `ContainerBuilder` and somehow... _transfer_ the registrations from one to another. It doesn't work like that. We couldn't just return a `ContainerBuilder` and then in the `CreateServiceProvider` step just "apply" those registrations to the scope.

If we could, we'd still need to figure out how `CreateServiceProvider` changes. It can't just call `ContainerBuilder.Build()` anymore, it'd have to decide whether it creates a lifetime scope (and, possibly, tags it) or whether it creates the container... and _then_ wraps the outcome in an `AutofacServiceProvider`.

I see there's a notion of a [base startup class](https://github.com/aspnet/AspNetCore/blob/master/src/Hosting/Hosting/src/Startup/StartupBase.cs) that has a way to manually return an `IServiceProvider` but I'm not sure if that's still a valid thing or not. In general the pattern isn't to derive from that anyway at the moment, so we may need to try that and see if it's an option for workaround... but it won't help the folks that can't change their startup classes to derive from that.

I also noticed in the PR and issue I linked earlier in the ASP.NET Core repo that one of the big things they are trying to force is the use of a single, global container. No double-containers, no partitions. One app, one container. You can't even inject a logger into `Startup` anymore - you get `IConfiguration` and `IHostingEnvironment` and _that's it_. They're intentionally making it hard for you to control these things.

After diving in here, and given the notes in the ASP.NET Core repo... while we may be able to do some pretty intricate workarounds to make the "ASP.NET Core app gets a child scope in a container instead of a whole container" thing... we may also be really trying to force a square peg into a round hole. I'm not sure how supportable it will be or if the changes we'd make to allow that to work would continue working in the next version when things get locked down even more.

If the `StartupBase` thing doesn't effectively work around it by allowing you to continue returning your own `IServiceProvider` out of `ConfigureServices` then individual projects/apps may want to create their own custom `IServiceProviderFactory<TContainerBuilder>` implementations that work enough to get this happening in that project but aren't something the Autofac team has to try to keep working across ASP.NET Core releases. It's already kind of hard to make sure they don't add things to their conforming container spec that don't work with... well, honestly, _most_ IoC containers. Keeping the 90% case working here and calling it good may be the most maintainable option.

And, of course, if someone does solve this in a maintainable way, we can take a PR or that person could publish their factory as their own NuGet package.

**Comment by MonkSoul** [comment]
I've got it done. https://www.cnblogs.com/baiqian/p/10964972.html

It's very simple.

**Comment by JuergenGutsch** [comment]
Just to get a conclusion out of that issue: There is currently no easy way to get Autofac running in ASP.NET Core 3.0. Is that right?

**Comment by tillig** [comment]
Don't know, haven't done work on it yet. Interesting using the service provider factory should just work, though I can't tell you the gotchas or workarounds. The problem I see mainly is that we likely can't continue to support hosting where the "root lifetime scope" for the hosted app _isn't_ a container. Some folks try to share a container and host their app with the root lifetime being a child scope. I don't think we'll be able to keep that.

**Comment by JuergenGutsch** [comment]
Ok. I'm going to try to use the service provider factory. (This issue kinda blocks me finishing a book about ASP.NET Core 3.0)
Did you try to talk to @davidfowl about that. He helped me with an issue I had with the [LightCore](https://github.com/juergengutsch/LightCore) IoC Container.

**Comment by JuergenGutsch** [comment]
As you wrote, using the service provider factory just works.

~~~ 
public class AutofacServiceProviderFactory : IServiceProviderFactory<ContainerBuilder>
{
    public ContainerBuilder CreateBuilder(IServiceCollection services)
    {
        var containerBuilder = new ContainerBuilder();

        // read service collection to Autofac
        containerBuilder.Populate(services);

        return containerBuilder;
    }

    public IServiceProvider CreateServiceProvider(ContainerBuilder containerBuilder)
    {
        var applicationContainer = containerBuilder.Build();

        // creating the IServiceProvider out of the Autofac container
        return new AutofacServiceProvider(applicationContainer);
    }
}
~~~

Usage:
~~~ 
public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        .UseServiceProviderFactory<ContainerBuilder>(new AutofacServiceProviderFactory()) // should be nicer
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder                       
                .UseStartup<Startup>();
        });
~~~

**Comment by davidfowl** [comment]
Autofac already has an AutofacServiceProviderFactory implementation you can use, no need to write your own.

> The problem I see mainly is that we likely can't continue to support hosting where the "root lifetime scope" for the hosted app isn't a container. Some folks try to share a container and host their app with the root lifetime being a child scope. I don't think we'll be able to keep that.

Can't that be a feature of the factory?

**Comment by JuergenGutsch** [comment]
> Autofac already has an AutofacServiceProviderFactory implementation you can use, no need to write your own.
Thanks @davidfowl :-) Yes I know. I did just to simply debug into things. Wasn't that much effort ;-)

**Comment by tillig** [comment]
There are a few challenges with trying to wrap things into the service provider factory that I'm not sure how to address. For the 80% case, the answer is:

- Remove the `ServiceCollectionExtensions.AddAutofac` extension - having the service provider factory inside the service collection doesn't work anymore. (Pretty sure.)
- Create a new `HostBuilderExtensions.UseAutofac` extension that attaches the Autofac service provider factory to the builder.
- Profit.

However, as I mentioned in my earlier fairly-long-and-rambling post, I don't know what that looks like if you _already have a container_ and the service provider factory is supposed to contribute to a child lifetime scope.

Let's say I have that container.

```c#
var builder = new ContainerBuilder();
var container = builder.Build();
```

When you create a lifetime scope, it's like this:

```c#
builder.BeginLifetimeScope("optional-tag", scopeBuilder => scopeBuilder.RegisterThingsHere());
```

The host builder startup process does this:

- Call `Startup.ConfigureServices` to get an `IServiceCollection`
- Call `AutofacServiceProviderFactory.CreateBuilder` to put the `IServiceCollection` into a `ContainerBuilder` and return that `ContainerBuilder`.
- Pass the `ContainerBuilder` to `Startup.ConfigureContainer` to register more stuff.
- Pass the `ContainerBuilder` to `AutofacServiceProviderFactory.CreateServiceProvider` to get an `IServiceProvider` with everything ready to go.

The problem is, `BeginLifetimeScope` executes that lambda immediately. It's not a deferred thing. We also don't have the notion of "run a bunch of registrations on one `ContainerBuilder` and apply those registrations against a second `ContainerBuilder`."

Here's the current `AutofacServiceProviderFactory`:

```c#
public class AutofacServiceProviderFactory : IServiceProviderFactory<ContainerBuilder>
{
  private readonly Action<ContainerBuilder> _configurationAction;

  public AutofacServiceProviderFactory(Action<ContainerBuilder> configurationAction = null)
  {
    _configurationAction = configurationAction ?? (builder => { });
  }

  public ContainerBuilder CreateBuilder(IServiceCollection services)
  {
    var builder = new ContainerBuilder();
    builder.Populate(services);
    _configurationAction(builder);
    return builder;
  }

  public IServiceProvider CreateServiceProvider(ContainerBuilder containerBuilder)
  {
    if (containerBuilder == null) throw new ArgumentNullException(nameof(containerBuilder));
    var container = containerBuilder.Build();
    return new AutofacServiceProvider(container);
  }
}
```

Let's say we were trying to support child lifetime scopes as a root for the app. How do I split the `BeginLifetimeScope` lambda across `CreateBuilder` and `CreateServiceProvider`? With the stuff we have at the moment, the answer to that isn't coming to me.

I can imagine some sort of really not-pretty indirection, like maybe this:

```c#
public class AutofacServiceProviderFactory : IServiceProviderFactory<IList<Action<ContainerBuilder>>>
{
  private readonly Action<ContainerBuilder> _configurationAction;
  private readonly ILifetimeScope _parent;

  public AutofacServiceProviderFactory(ILifetimeScope parent, Action<ContainerBuilder> configurationAction = null)
  {
    _parent = parent;
    _configurationAction = configurationAction ?? (builder => { });
  }

  public IList<Action<ContainerBuilder>> CreateBuilder(IServiceCollection services)
  {
    var actions = new List<Action<ContainerBuilder>>();
    actions.Add(b => b.Populate(services));
    actions.Add(b => _configurationAction(b));
    return actions;
  }

  public IServiceProvider CreateServiceProvider(IList<Action<ContainerBuilder>> containerBuilderActions)
  {
    if (containerBuilderActions == null) throw new ArgumentNullException(nameof(containerBuilderActions));
    var scope = _parent.BeginLifetimeScope(scopeBuilder =>
    {
      foreach(var action in containerBuilderActions)
      {
        action(scopeBuilder);
      }
    }
    return new AutofacServiceProvider(scope);
  }
}
```

The usage on that would sort of be sucky in the `Startup` but maybe that's the price you pay for wanting to be in a child scope.

```c#
public void ConfigureContainer(IList<Action<ContainerBuilder>> actions)
{
  actions.Add(b =>
  {
    b.RegisterType<Foo>().As<IFoo>());
    b.RegisterType<Bar>().As<IBar>());
  });
}
```

Likely we'd make that a different provider factory so the 80% case would be easier.

We have a similar challenge with multitenant support in the new world. Building a multitenant container is like:

```c#
var builder = new ContainerBuilder();
var container = builder.Build();
var multitenantContainer = new MultitenantContainer(container, new TenantIdentificationStrategy());
multitenantContainer.RegisterTenant("tenant-1", b =>
{
  b.Register<TenantDependency>();
});
```

In the situation where you could return the `IServiceProvider` from `ConfigureServices` this was an easier thing to handle, or at least more straightforward to understand. Forcing it through a factory, I'm not sure how to deal with it. Maybe there's a multitenant factory that takes a delegate to do the multitenant configuration after the fact? I don't get a reference in the factory to the startup class, so I can't make that interaction automatic or pretty. That also means I can't use an instance method to get things like configuration or hosting environment that I'd potentially otherwise have in a startup class.

Point being, unless I'm missing something... the 80% case is easy to solve, but these other use cases don't feel so straightforward in the new world.

**Comment by tillig** [comment]
I'm also not sure how to address the issue where folks want to keep their own reference to the container. If the container is built _inside the factory_ and the factory returns an `IServiceProvider` then there's no room to return _the container itself, too_. Options I can think of off the top of my head:

- Allow a lambda to be passed to the factory that takes a container (or lifetime scope). The factory can call that lambda after the container is built. In that lambda, you can set a static variable to the container.
- Folks can register [a container build callback](https://autofaccn.readthedocs.io/en/latest/lifetime/startup.html#container-build-callbacks) that will set the global variable to the container reference. This isn't a thing on lifetime scopes, though, IIRC, so the folks wanting child lifetime scopes as the "root" for the app may be stuck. I'd have to test it.
- Run some logic that resolves `ILifetimeScope` from the `IServiceProvider` - this will give the lifetime scope in the service provider. Store that. This is probably the best option, but won't work in multitenant mode because all resolutions from the multitenant container factor in the tenant. To properly support multitenant request services, we absolutely need the reference to the root container.

**Comment by tillig** [comment]
More brainstorming: Perhaps an `IServiceProvider` extension that tries to cast the provider to `AutofacServiceProvider` and get container (or whatever) references that way. We'd have to add a property to the `AutofacServiceProvider` to get the root lifetime scope, but it could be something like:

```c#
public ILifetimeScope GetAutofacRoot(this IServiceProvider provider)
{
  return (provider as AutofacServiceProvider)?.LifetimeScope;
}
```

That would solve the issue of folks wanting the reference to the container and would keep it out of the factory.

**Comment by davidfowl** [comment]
> More brainstorming: Perhaps an IServiceProvider extension that tries to cast the provider to AutofacServiceProvider and get container (or whatever) references that way. We'd have to add a property to the AutofacServiceProvider to get the root lifetime scope, but it could be something like:

I was going to suggest that. It's a much better approach as it makes the autofac ILifetimeScope available everywhere.

As for the other thing, I'm a fan of the indirection but maybe not using `IList<Action<ContainerBuilder>>`, maybe it could be some wrapper type that doesn't look as noisy for these more advanced scenarios?

**Comment by tillig** [comment]
Yeah, we could have like `public class ScopeBuilder : List<Action<ContainerBuilder>> { }` or something to make it a little nicer.

Not totally sure about the multitenant container thing, possibly solvable with a similar strategy. Most of this is going to boil down to getting the time [and, honestly, motivation... I've a non-zero level of OSS maintainer burnout set in] to sit down, work up a few options, try them out, see how they work, and get it done.

**Comment by davidfowl** [comment]
You could also consider having a single builder that can handle all scenarios but I haven't looked at the API deeply enough to know how trivial it would be.

**Comment by netleader** [comment]
Hi Folks,
any news on this topic? I'm fighting to get my projects moved to 3.0 with MultitenantContainer support but due to the breaking changes I can't get this to work.

**Comment by alsami** [comment]
Hi! We are working on it and we will keep this issue up to that.

**Comment by alsami** [comment]
@autofac/contributors I have create a small [repository](https://github.com/alsami/AutofacNetcore3) containing libraries and test projects that need changing for .NET Core 3 / ASP.NET Core 3 support. I copied those from the original repository since I am still having a hard time understanding all the problems at hand. I think this way it might be easier to test during development.

I had to remove the references to full .NET-Framework for now since I am running Linux, which shouldn't be a problem for now. Later I will create virtual-machine. Right now .NET-Core 2.2 and 3.0-Preview-7 are required to build the solution.

I created a small recap of the problems and todos [here](https://github.com/alsami/AutofacNetcore3/blob/master/issue_36.md).

If anyone wants to jump on board and help out I'd be grateful :smile:

**Comment by alsami** [comment]
> More brainstorming: Perhaps an `IServiceProvider` extension that tries to cast the provider to `AutofacServiceProvider` and get container (or whatever) references that way. We'd have to add a property to the `AutofacServiceProvider` to get the root lifetime scope, but it could be something like:
> 
> ```cs
> public ILifetimeScope GetAutofacRoot(this IServiceProvider provider)
> {
>   return (provider as AutofacServiceProvider)?.LifetimeScope;
> }
> ```
> 
> That would solve the issue of folks wanting the reference to the container and would keep it out of the factory.

I have implemented an extension for this in the library `Autofac.Extensions.DependencyInjection` with this [commit](https://github.com/alsami/AutofacNetcore3/commit/ed05b956f56e6506169ed91ec732f7bb4f99406a) within the test-repository.

I decided to check whether the underlying implementation of `IServiceProvider` is in fact of type `AutofacServiceProvider` and wrote tests for that as well. I guess we can still remove it when it's not required but I thought it's better than returning `null`.

I've also implemented a prototype for `AutofacRootScopeServiceProviderFactory` (I wasn't sure about how to name it) quite similar to what you @davidfowl  suggested by using a [wrapper](https://github.com/alsami/AutofacNetcore3/blob/0023dc81177cbb9c034905673d9797f93fc6b428/src/Autofac.Extensions.DependencyInjection/AutofacRootScopeServiceProviderFactoryActions.cs#L13) for the actions that need to applied on each scope.

While I have implemented and tested it manually [here](https://github.com/alsami/AutofacNetcore3/blob/0023dc81177cbb9c034905673d9797f93fc6b428/src/AutofacNetcore3.Rootscope/Startup.cs#L29) and I think that it is working, I am not sure if that's exactly what is needed.

Would you elaborate a little more what are the use-cases for this exactly and when would I want my container to only use child-scopes?

Some feedback and input would be highly appreciated.

**Comment by alsami** [comment]
While migrating one of my service to ASP.NET Core 3 I noticed that it's really not necessary to use the new generic-host-builder. I don't know if that will change in the future, but everything is still working as before when using the previous valid syntax to build a web-host.

This would not require to deprecate the usage of `ConfigureServices(sp => sp.AddAutofac());`

@davidfowl can you tell us if it will be required to use the generic-host with `IHostBuilder` in the final version of ASP.NET Core 3?

**Comment by tillig** [comment]
**We need to support the generic hosting mechanism.** This extension library is not just about ASP.NET Core, though that is a prime use case.

[This pull request](https://github.com/autofac/Autofac.Extensions.DependencyInjection/pull/14) shows the context for the "use a nested lifetime scope" use case. Some folks want to host multiple application types in a single app, so the actual root container is shared across, say, ASP.NET Core and other components, but the ASP.NET Core stuff is _only registered in a child lifetime scope_ and that child scope serves as the "root" for the ASP.NET Core part of the app.

The multitenant use cases were initially very specific to ASP.NET Core [since they require interaction with request services](https://github.com/autofac/Autofac.AspNetCore.Multitenant). I don't know what changes will have to be made here to accommodate for multitenant support. If it turns out we need a special "multitenant service provider factory" or something like that, we'll probably want a non-ASP.NET-Core library for that (e.g., "Autofac.Extensions.DependencyInjection.Multitenant" or something). I am not _recommending this_ I'm just saying _it's possible_. I haven't done the leg work to see if this is a thing. It may be equally possible that all the changes can be made right here in this library and it'll "just work" for multitenancy. (We definitely don't want any Autofac.Multitenant references in this library, though, so as soon as that appears required, it'll call for a split.)

I will need some help from @alexmg  and/or @nblumhardt  to push this through. I have some personal issues I'm working through right now and don't have time to dedicate here on a consistent basis. I won't be able to respond or review things in a timely fashion.

**Comment by alsami** [comment]
> **We need to support the generic hosting mechanism.** This extension library is not just about ASP.NET Core, though that is a prime use case.

Yes, I know we have to. I just stated that as of now, ASP.NET Core 3 works fine with 'just' a web-host and that the current extension for it still works (which of course might change).

> [This pull request](https://github.com/autofac/Autofac.Extensions.DependencyInjection/pull/14) shows the context for the "use a nested lifetime scope" use case. Some folks want to host multiple application types in a single app, so the actual root container is shared across, say, ASP.NET Core and other components, but the ASP.NET Core stuff is _only registered in a child lifetime scope_ and that child scope serves as the "root" for the ASP.NET Core part of the app.

Thanks for the explanation, now I get it. What you think about the prototype I've implemented? Seeing any problems with it?

> I will need some help from @alexmg and/or @nblumhardt to push this through.

That would be great, currently I can't do it on my own :) 

> I have some personal issues I'm working through right now and don't have time to dedicate here on a consistent basis. I won't be able to respond or review things in a timely fashion.

Hope things will be better soon, hanks for the feedback!

**Comment by alsami** [comment]
> Hi Folks,
> any news on this topic? I'm fighting to get my projects moved to 3.0 with MultitenantContainer support but due to the breaking changes I can't get this to work.

Hey, they changes required for it have been merged with this [PR](https://github.com/autofac/Autofac.AspNetCore.Multitenant/pull/10). 

Just wanted to let you know, there is currently no release for it. If you want to use it already, feel free to copy the code for now.

**Comment by tillig** [comment]
I will likely release a beta package shortly. Also, don't forget there's a MyGet feed for early adopters where you can pick up and use any of our CI builds: `https://www.myget.org/F/autofac/`

**Comment by tillig** [comment]
[Docs are adjusted](https://autofac.readthedocs.io/en/latest/integration/aspnetcore.html) and an RC package just went out. This will be the 5.0.0 release when it's finalized, but we'll wait on that final release until after .NET Core 3 itself is released.
